test/blackbox-tests/test-cases/actions/case.t

```diff
@@ -0,0 +1,149 @@
+Testing (case) action.
+
+  $ cat > dune-project << EOF
+  > (lang dune 3.11)
+  > EOF
+
+  $ cat > version << EOF
+  > 1
+  > EOF
+
+Basic usage:
+
+  $ cat > dune << EOF
+  > (rule
+  >  (alias foo)
+  >  (action
+  >   (case %{read-lines:version}
+  >    (1 (echo "version is 1"))
+  >    (2 (echo "version is 2"))
+  >    (3 (echo "version is 3"))
+  >    (_ (echo "shouldn't happen, got %{read-lines:version}")))))
+  > EOF
+
+  $ dune build @foo
+  version is 1
+
+  $ cat > version << EOF
+  > 2
+  > EOF
+  $ dune build @foo
+  version is 2
+
+  $ cat > version << EOF
+  > 3
+  > EOF
+  $ dune build @foo
+  version is 3
+
+  $ cat > version << EOF
+  > 4
+  > EOF
+  $ dune build @foo
+  shouldn't happen, got 4
+
+  $ cat > version << EOF
+  > 1
+  > EOF
+
+Missing default field:
+
+  $ cat > dune << EOF
+  > (rule
+  >  (alias foo)
+  >  (action
+  >   (case %{read-lines:version}
+  >    (1 (echo "version is 1")))))
+  > EOF
+
+  $ dune build @foo
+  File "dune", line 5, characters 4-5:
+  5 |    (1 (echo "version is 1")))))
+          ^
+  Error: Only the default case can be at the end.
+  Hint: Add a (_ (...)) case at the end.
+  [1]
+
+Two default fields:
+
+  $ cat > dune << EOF
+  > (rule
+  >  (alias foo)
+  >  (action
+  >   (case %{read-lines:version}
+  >    (_ (echo "version is 1"))
```

Comment:
Strings in dune can be unquoted. What happens if the `version` file contains "_"? Would the `_` case be treated as the default case or matching the "_" string?

What if we changed the syntax so the default case doesn't require a condition at all. E.g.
```
(case %{read-lines:version}
 (2 (echo "version is 2"))
 (echo "version is not 2"))
```

Reply:
If `version` contained `_` it would be compared to any other branch with pattern `_`. You can give such a pattern by quoting the `_`. For instance, here is a test I have added:

```

  $ cat > dune << EOF
  > (rule
  >  (alias foo)
  >  (action
  >   (case %{read-lines:version}
  >    ("_" (echo "version is _"))
  >    (_ (echo "version is default")))))
  > EOF

  $ dune build @foo
  version is default

  $ cat > version << EOF
  > _
  > EOF

  $ dune build @foo
  version is _

``` 

### You second point:

This makes parsing more complicated. In fact it can event be ambiguous. Imagine you had a file called `progn`. Then
```
(progn (echo foo))
```
could mean the default case or the string `progn`.

Reply:
Is the `_` string a special value in other places in dune? As a user I would find it confusing for there to be some values which are treated differently when they are quoted since strings without spaces don't usually need to be quoted in dune.

Reply:
It has a precedent as a catchall value in the env stanza.